File: app/src/main/java/com/waray/spendhound/PayorAdapter.java
package com.waray.spendhound;

import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class PayorAdapter extends RecyclerView.Adapter<PayorAdapter.PayorViewHolder> {

    // Static cache for download URLs to reduce Firebase Storage calls
    private static final Map<String, String> sDownloadUrlCache = new ConcurrentHashMap<>();

    private List<String> payorsUids;
    private List<String> payorsNames;
    private List<Double> amountsPaid;
    private List<Double> originalAmountsPaid;
    private double individualPayment;
    private OnPayorClickListener onPayorClickListener;
    private OnLoadingCompleteListener loadingCompleteListener;
    private OnDataChangedListener dataChangedListener;
    private Set<Integer> loadedPositions = new HashSet<>();
    private Set<Integer> failedPositions = new HashSet<>();
    private boolean isEditMode = false;

    public interface OnPayorClickListener {
        void onPayorClick(int index, double paid);
        void onPartialClick(int index, double currentPaid);
    }

    public interface OnLoadingCompleteListener {
        void onLoadingComplete();
    }

    public interface OnDataChangedListener {
        void onDataChanged(boolean hasChanges);
    }

    public void setOnLoadingCompleteListener(OnLoadingCompleteListener listener) {
        this.loadingCompleteListener = listener;
        if ((payorsUids == null || payorsUids.isEmpty()) && loadingCompleteListener != null) {
            loadingCompleteListener.onLoadingComplete();
        }
    }

    public void setOnDataChangedListener(OnDataChangedListener listener) {
        this.dataChangedListener = listener;
    }

    public PayorAdapter(List<String> payorsUids, List<String> payorsNames, List<Double> amountsPaid, double individualPayment, OnPayorClickListener onPayorClickListener) {
        this.payorsUids = payorsUids;
        this.payorsNames = payorsNames;
        this.amountsPaid = new ArrayList<>(amountsPaid);
        this.originalAmountsPaid = new ArrayList<>(amountsPaid);
        this.individualPayment = individualPayment;
        this.onPayorClickListener = onPayorClickListener;
    }

    public void setEditMode(boolean editMode) {
        this.isEditMode = editMode;
        if (!editMode) {
            // Revert changes if cancel or finishing edit
            this.amountsPaid = new ArrayList<>(originalAmountsPaid);
        }
        loadedPositions.clear();
        failedPositions.clear();
        notifyDataSetChanged();
        notifyDataChanged();
    }

    public void saveChanges() {
        this.originalAmountsPaid = new ArrayList<>(amountsPaid);
        this.isEditMode = false;
        loadedPositions.clear();
        failedPositions.clear();
        notifyDataSetChanged();
        notifyDataChanged();
    }

    public boolean hasChanges() {
        if (amountsPaid.size() != originalAmountsPaid.size()) return true;
        for (int i = 0; i < amountsPaid.size(); i++) {
            if (!amountsPaid.get(i).equals(originalAmountsPaid.get(i))) {
                return true;
            }
        }
        return false;
    }

    private void notifyDataChanged() {
        if (dataChangedListener != null) {
            dataChangedListener.onDataChanged(hasChanges());
        }
    }

    public List<Double> getAmountsPaid() {
        return amountsPaid;
    }

    @NonNull
    @Override
    public PayorViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_payor_horizontal, parent, false);
        return new PayorViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PayorViewHolder holder, int position) {
        String uid = payorsUids.get(position);
        String name = (payorsNames != null && position < payorsNames.size()) ? payorsNames.get(position) : "User";
        double paid = (amountsPaid != null && position < amountsPaid.size()) ? amountsPaid.get(position) : 0.0;

        holder.payorName.setText(name);
        holder.payorPayment.setText(CurrencyUtils.formatAmount(paid) + "/" + CurrencyUtils.formatAmount(individualPayment));

        updateStatusUI(holder, paid);

        if (isEditMode) {
            holder.editButtonsLayout.setVisibility(View.VISIBLE);
            
            // Logic to hide buttons based on status
            if (paid <= 0) {
                holder.unpaidBtn.setVisibility(View.GONE);
                holder.paidBtn.setVisibility(View.VISIBLE);
                holder.partialBtn.setVisibility(View.VISIBLE);
            } else if (paid >= individualPayment) {
                holder.unpaidBtn.setVisibility(View.VISIBLE);
                holder.paidBtn.setVisibility(View.GONE);
                holder.partialBtn.setVisibility(View.VISIBLE);
            } else {
                holder.unpaidBtn.setVisibility(View.VISIBLE);
                holder.paidBtn.setVisibility(View.VISIBLE);
                holder.partialBtn.setVisibility(View.VISIBLE);
            }

            holder.unpaidBtn.setOnClickListener(v -> {
                amountsPaid.set(position, 0.0);
                notifyItemChanged(position);
                notifyDataChanged();
            });

            holder.paidBtn.setOnClickListener(v -> {
                amountsPaid.set(position, individualPayment);
                notifyItemChanged(position);
                notifyDataChanged();
            });

            holder.partialBtn.setOnClickListener(v -> {
                if (onPayorClickListener != null) {
                    onPayorClickListener.onPartialClick(position, amountsPaid.get(position));
                }
            });
        } else {
            holder.editButtonsLayout.setVisibility(View.GONE);
        }

        // Use cached download URL if available
        String cachedUrl = sDownloadUrlCache.get(uid);
        if (cachedUrl != null) {
            loadGlideImage(holder, cachedUrl, position);
        } else {
            StorageReference pStorageRef = FirebaseStorage.getInstance().getReference("profile_images").child(uid);
            pStorageRef.getDownloadUrl().addOnSuccessListener(uri -> {
                String url = uri.toString();
                sDownloadUrlCache.put(uid, url);
                loadGlideImage(holder, url, position);
            }).addOnFailureListener(e -> {
                holder.payorImage.setImageResource(R.drawable.placeholder_profile_image);
                checkLoadingComplete(position);
            });
        }

        if (!isEditMode && onPayorClickListener != null) {
            holder.itemView.setOnClickListener(v -> onPayorClickListener.onPayorClick(position, paid));
        } else {
            holder.itemView.setOnClickListener(null);
        }
    }

    private void loadGlideImage(PayorViewHolder holder, String url, int position) {
        Glide.with(holder.itemView.getContext())
                .load(url)
                .placeholder(R.drawable.placeholder_profile_image)
                .diskCacheStrategy(DiskCacheStrategy.ALL) // Enable full disk caching
                .circleCrop()
                .listener(new RequestListener<Drawable>() {
                    @Override
                    public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                        checkLoadingComplete(position);
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                        checkLoadingComplete(position);
                        return false;
                    }
                })
                .into(holder.payorImage);
    }

    private void updateStatusUI(PayorViewHolder holder, double paid) {
        if (paid <= 0) {
            holder.payorStatus.setText("Unpaid");
            holder.payorStatus.setTextColor(holder.itemView.getContext().getResources().getColor(android.R.color.holo_red_dark));
        } else if (paid < individualPayment) {
            holder.payorStatus.setText("Paid Partially");
            holder.payorStatus.setTextColor(holder.itemView.getContext().getResources().getColor(android.R.color.holo_orange_dark));
        } else {
            holder.payorStatus.setText("Paid");
            holder.payorStatus.setTextColor(holder.itemView.getContext().getResources().getColor(android.R.color.holo_green_dark));
        }
    }

    public void updatePartialAmount(int index, double amount) {
        if (index < amountsPaid.size()) {
            amountsPaid.set(index, amount);
            notifyItemChanged(index);
            notifyDataChanged();
        }
    }

    private synchronized void checkLoadingComplete(int position) {
        loadedPositions.add(position);
        // We consider it complete if all items have at least attempted to load
        if (loadedPositions.size() >= getItemCount() && loadingCompleteListener != null) {
            loadingCompleteListener.onLoadingComplete();
        }
    }

    @Override
    public int getItemCount() {
        return payorsUids != null ? payorsUids.size() : 0;
    }

    static class PayorViewHolder extends RecyclerView.ViewHolder {
        ImageView payorImage;
        TextView payorName, payorPayment, payorStatus;
        View editButtonsLayout;
        Button unpaidBtn, paidBtn, partialBtn;

        public PayorViewHolder(@NonNull View itemView) {
            super(itemView);
            payorImage = itemView.findViewById(R.id.payorProfileImage);
            payorName = itemView.findViewById(R.id.payorNameTextView);
            payorPayment = itemView.findViewById(R.id.payorPaymentTextView);
            payorStatus = itemView.findViewById(R.id.payorStatusTextView);
            editButtonsLayout = itemView.findViewById(R.id.editButtonsLayout);
            unpaidBtn = itemView.findViewById(R.id.unpaid_btn);
            paidBtn = itemView.findViewById(R.id.paid_btn);
            partialBtn = itemView.findViewById(R.id.partial_btn);
        }
    }
}

File: app/src/main/java/com/waray/spendhound/PayorAdapter.kt
package com.waray.spendhound

import android.content.Context
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import coil.imageLoader
import coil.load
import coil.transform.CircleCropTransformation
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap

class PayorAdapter(
    private val payorUserIds: MutableList<String?>?,
    private val payorsNames: MutableList<String?>?,
    amountsPaid: MutableList<Double?>,
    private val individualPayments: List<Double>,
    private val excessAmounts: List<Double>,
    private val onPayorClickListener: OnPayorClickListener?
) : RecyclerView.Adapter<PayorAdapter.PayorViewHolder>() {
    var amountsPaid: MutableList<Double?>? = null
        private set
    private var originalAmountsPaid: MutableList<Double?>
    private var loadingCompleteListener: OnLoadingCompleteListener? = null
    private var dataChangedListener: OnDataChangedListener? = null
    private val loadedPositions: MutableSet<Int> = HashSet()
    private var isEditMode = false
    private val scope = CoroutineScope(Dispatchers.Main)

    interface OnPayorClickListener {
        fun onPayorClick(index: Int, paid: Double)
        fun onPartialClick(index: Int, currentPaid: Double)
    }

    interface OnLoadingCompleteListener {
        fun onLoadingComplete()
    }

    interface OnDataChangedListener {
        fun onDataChanged(hasChanges: Boolean)
    }

    fun setOnLoadingCompleteListener(listener: OnLoadingCompleteListener?) {
        this.loadingCompleteListener = listener
        if (payorUserIds.isNullOrEmpty() && loadingCompleteListener != null) {
            loadingCompleteListener!!.onLoadingComplete()
        }
    }

    fun setOnDataChangedListener(listener: OnDataChangedListener?) {
        this.dataChangedListener = listener
    }

    init {
        this.amountsPaid = ArrayList(amountsPaid)
        this.originalAmountsPaid = ArrayList(amountsPaid)
    }

    fun startLoadingAllImages(context: Context) {
        if (payorUserIds.isNullOrEmpty()) {
            loadingCompleteListener?.onLoadingComplete()
            return
        }

        for (i in payorUserIds.indices) {
            val userId = payorUserIds[i] ?: continue
            val cachedUrl = sDownloadUrlCache[userId]

            if (cachedUrl != null) {
                preloadProfileImage(context, cachedUrl, i)
            } else {
                scope.launch {
                    try {
                        val uidLong = userId.toLongOrNull()
                        val user = withContext(Dispatchers.IO) {
                            if (uidLong != null) {
                                DeclareDatabase.usersTable.select { filter { eq("user_id", uidLong) } }.decodeSingleOrNull<User>()
                            } else null
                        }
                        val url = user?.profileImageUrl?.takeIf { it.isNotBlank() && it != "placeholder_profile_image" }
                            ?: withContext(Dispatchers.IO) {
                                DeclareDatabase.profileImagesBucket.publicUrl("$userId/$userId.jpg")
                            }
                        sDownloadUrlCache[userId] = url
                        preloadProfileImage(context, url, i)
                    } catch (e: Exception) {
                        checkLoadingComplete(i)
                    }
                }
            }
        }
    }

    private fun preloadProfileImage(context: Context, url: String, position: Int) {
        // Coil preloading
        val request = coil.request.ImageRequest.Builder(context)
            .data(url)
            .transformations(CircleCropTransformation())
            .listener(
                onSuccess = { _, _ -> checkLoadingComplete(position) },
                onError = { _, _ -> checkLoadingComplete(position) }
            )
            .build()
        context.imageLoader.enqueue(request)
    }

    fun setEditMode(editMode: Boolean) {
        this.isEditMode = editMode
        if (!editMode) {
            this.amountsPaid = ArrayList(originalAmountsPaid)
        }
        loadedPositions.clear()
        notifyDataSetChanged()
        notifyDataChanged()
    }

    fun saveChanges() {
        this.originalAmountsPaid = ArrayList(amountsPaid!!)
        this.isEditMode = false
        loadedPositions.clear()
        notifyDataSetChanged()
        notifyDataChanged()
    }

    fun hasChanges(): Boolean {
        if (amountsPaid!!.size != originalAmountsPaid.size) return true
        for (i in amountsPaid!!.indices) {
            if (amountsPaid!![i] != originalAmountsPaid[i]) return true
        }
        return false
    }

    private fun notifyDataChanged() {
        dataChangedListener?.onDataChanged(hasChanges())
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PayorViewHolder {
        val view: View = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_payor_horizontal, parent, false)
        return PayorViewHolder(view)
    }

    override fun onBindViewHolder(holder: PayorViewHolder, position: Int) {
        val userId = payorUserIds?.get(position)
        val name = if (payorsNames != null && position < payorsNames.size) payorsNames[position] ?: "User" else "User"
        val paid = if (amountsPaid != null && position < amountsPaid!!.size) amountsPaid!![position] ?: 0.0 else 0.0
        val individualPayment = individualPayments.getOrNull(position) ?: 0.0

        holder.payorName.text = name

        updateStatusUI(holder, paid, position)

        val diff = paid - individualPayment
        val excess = excessAmounts.getOrNull(position) ?: 0.0
        val epsilon = 0.01

        // Creditors show current excess (remaining receivable), debtors show diff (remaining debt)
        val balance = if (diff > epsilon) excess else diff

        if (kotlin.math.abs(balance) > epsilon) {
            holder.payorBalanceAmount.visibility = View.VISIBLE
            val sign = if (balance > epsilon) "+" else ""
            holder.payorBalanceAmount.text = "$sign${CurrencyUtils.formatAmountWithCurrency(balance)}"
            holder.payorBalanceAmount.setTextColor(ContextCompat.getColor(holder.itemView.context, if (balance > epsilon) R.color.green else R.color.red))
        } else {
            holder.payorBalanceAmount.visibility = View.GONE
        }

        if (isEditMode) {
            holder.editButtonsLayout.visibility = View.VISIBLE
            holder.unpaidBtn.visibility = if (paid > 0) View.VISIBLE else View.GONE
            holder.paidBtn.visibility = if (paid < individualPayment) View.VISIBLE else View.GONE
            holder.partialBtn.visibility = View.VISIBLE

            holder.unpaidBtn.setOnClickListener {
                amountsPaid!![position] = 0.0
                notifyItemChanged(position)
                notifyDataChanged()
            }

            holder.paidBtn.setOnClickListener {
                amountsPaid!![position] = individualPayment
                notifyItemChanged(position)
                notifyDataChanged()
            }

            holder.partialBtn.setOnClickListener {
                onPayorClickListener?.onPartialClick(position, amountsPaid!![position]!!)
            }
        } else {
            holder.editButtonsLayout.visibility = View.GONE
        }

        if (userId != null) {
            val cachedUrl = sDownloadUrlCache[userId] ?: UserHelper.getCachedImageUrl(userId.toLongOrNull())
            if (cachedUrl != null) {
                loadCoilImage(holder, cachedUrl, position, name)
            } else {
                scope.launch {
                    try {
                        val uidLong = userId.toLongOrNull()
                        val user = withContext(Dispatchers.IO) {
                            if (uidLong != null) {
                                DeclareDatabase.usersTable.select { filter { eq("user_id", uidLong) } }.decodeSingleOrNull<User>()
                            } else null
                        }
                        val url = user?.profileImageUrl?.takeIf { it.isNotBlank() && it != "placeholder_profile_image" }
                            ?: withContext(Dispatchers.IO) {
                                DeclareDatabase.profileImagesBucket.publicUrl("$userId/$userId.jpg")
                            }
                        sDownloadUrlCache[userId] = url
                        if (uidLong != null) {
                            user?.username?.let { UserHelper.updateCache(uidLong, it, url) }
                        }
                        loadCoilImage(holder, url, position, name)
                    } catch (e: Exception) {
                        holder.payorInitials.text = name.take(2).uppercase()
                        holder.payorInitials.visibility = View.VISIBLE
                        holder.payorImage.visibility = View.GONE
                        checkLoadingComplete(position)
                    }
                }
            }
        }

        if (!isEditMode) {
            holder.itemView.setOnClickListener { onPayorClickListener?.onPayorClick(position, paid) }
        } else {
            holder.itemView.setOnClickListener(null)
        }
    }

    private fun loadCoilImage(holder: PayorViewHolder, url: String?, position: Int, name: String) {
        holder.payorInitials.text = name.take(2).uppercase()
        holder.payorInitials.visibility = View.VISIBLE
        
        // Ensure image view is visible for Coil to load into it, but clear it first
        holder.payorImage.visibility = View.VISIBLE
        holder.payorImage.setImageDrawable(null)
        holder.payorImage.imageTintList = null

        if (url.isNullOrBlank()) {
            holder.payorImage.visibility = View.GONE
            checkLoadingComplete(position)
            return
        }

        holder.payorImage.load(url) {
            transformations(CircleCropTransformation())
            crossfade(true)
            listener(
                onSuccess = { _, _ ->
                    holder.payorInitials.visibility = View.GONE
                    holder.payorImage.visibility = View.VISIBLE
                    checkLoadingComplete(position)
                },
                onError = { _, _ ->
                    holder.payorInitials.visibility = View.VISIBLE
                    holder.payorImage.visibility = View.GONE
                    checkLoadingComplete(position)
                }
            )
        }
    }

    private fun updateStatusUI(holder: PayorViewHolder, paid: Double, position: Int) {
        val context = holder.itemView.context
        val individualPayment = individualPayments.getOrNull(position) ?: 0.0
        when {
            paid <= 0 -> {
                holder.payorStatus.text = "Unpaid"
                holder.payorStatus.setTextColor(ContextCompat.getColor(context, R.color.bright_red))
                holder.payorStatusBadge.visibility = View.VISIBLE
                holder.payorStatusBadge.setImageResource(R.drawable.ic_circle_checked)
                holder.payorStatusBadge.setColorFilter(ContextCompat.getColor(context, R.color.bright_red))
            }
            paid < individualPayment -> {
                holder.payorStatus.text = "Paid Partially"
                holder.payorStatus.setTextColor(ContextCompat.getColor(context, R.color.mid_orange))
                holder.payorStatusBadge.visibility = View.VISIBLE
                holder.payorStatusBadge.setImageResource(R.drawable.ic_circle_checked)
                holder.payorStatusBadge.setColorFilter(ContextCompat.getColor(context, R.color.mid_orange))
            }
            else -> {
                holder.payorStatus.text = "Paid"
                holder.payorStatus.setTextColor(ContextCompat.getColor(context, R.color.green))
                holder.payorStatusBadge.visibility = View.VISIBLE
                holder.payorStatusBadge.setImageResource(R.drawable.ic_circle_checked)
                holder.payorStatusBadge.setColorFilter(ContextCompat.getColor(context, R.color.green))
            }
        }
    }

    fun updatePartialAmount(index: Int, amount: Double) {
        if (index < amountsPaid!!.size) {
            amountsPaid!![index] = amount
            notifyItemChanged(index)
            notifyDataChanged()
        }
    }

    @Synchronized
    private fun checkLoadingComplete(position: Int) {
        loadedPositions.add(position)
        if (loadedPositions.size >= itemCount) {
            loadingCompleteListener?.onLoadingComplete()
        }
    }

    override fun getItemCount(): Int = payorUserIds?.size ?: 0

    class PayorViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val payorImage: ImageView = itemView.findViewById(R.id.payorProfileImage)
        val payorInitials: TextView = itemView.findViewById(R.id.payorInitialsTextView)
        val payorStatusBadge: ImageView = itemView.findViewById(R.id.payorStatusBadge)
        val payorName: TextView = itemView.findViewById(R.id.payorNameTextView)

        val payorStatus: TextView = itemView.findViewById(R.id.payorStatusTextView)
        val payorBalanceAmount: TextView = itemView.findViewById(R.id.payorBalanceAmountTextView)
        val editButtonsLayout: View = itemView.findViewById(R.id.editButtonsLayout)
        val unpaidBtn: Button = itemView.findViewById(R.id.unpaid_btn)
        val paidBtn: Button = itemView.findViewById(R.id.paid_btn)
        val partialBtn: Button = itemView.findViewById(R.id.partial_btn)
    }

    companion object {
        val sDownloadUrlCache: MutableMap<String, String> = ConcurrentHashMap()

        fun preCacheUserIds(context: Context, userIds: MutableList<String?>) {
            val scope = CoroutineScope(Dispatchers.IO)
            for (userId in userIds) {
                if (userId == null) continue
                if (sDownloadUrlCache.containsKey(userId)) continue
                scope.launch {
                    try {
                        val url = DeclareDatabase.profileImagesBucket.publicUrl("$userId/$userId.jpg")
                        sDownloadUrlCache[userId] = url
                    } catch (e: Exception) {}
                }
            }
        }

        private fun preloadOnly(context: Context, url: String) {
            val request = coil.request.ImageRequest.Builder(context)
                .data(url)
                .transformations(CircleCropTransformation())
                .build()
            context.imageLoader.enqueue(request)
        }
    }
}
